File: Frontend/angular-app/src/app/modules/admin/repositories/user/commands/update-user-command.ts
import {IRequest, IRequestHandler} from "../../../../../core/services/mediator/interfaces";
import {User} from "../../../entities/user";
import {ValidationRule} from "../../../../../core/validation/validation-rule";
import {MediatorHandler} from "../../../../../core/services/mediator/decorator";
import {Inject} from "@angular/core";
import {HttpService} from "../../../../../core/services/http/http.service";
import {HttpRequest} from "../../../../../core/services/http/http-request";
import {ServiceResult} from "../../../../../core/models/service-result";
import {NotificationService} from "../../../../../shared/services/notification/notification.service";

export class UpdateUserCommand extends IRequest<UpdateUserCommand, User>{
  public id:number | undefined = undefined;
  public personId:number | undefined = undefined;
  public username:string | undefined = undefined;
  public password:string | undefined = undefined;
  public confirmPassword:string | undefined = undefined;
  public isBlocked:boolean | undefined = undefined;
  public blockedReasonBaseId:boolean | undefined = undefined;
  public rolesIdList:number[] = [];
  public userAllowedYears:number[] = [];

  public lastOnlineTime:Date  | undefined = undefined;
  public failedCount:number  | undefined = undefined;
  public unitPositionTitle:string  | undefined = undefined;
  public firstName:string  | undefined = undefined;
  public lastName:string  | undefined = undefined;
  public nationalNumber:string  | undefined = undefined;

  mapFrom(entity: User): UpdateUserCommand {
    this.mapBasics(entity,this)
    this.rolesIdList = entity.rolesIdList
    this.userAllowedYears = entity.userYears
    return this;
  }

  mapTo(): User {
    return new User();
  }

  get url(): string {
    return "/admin/User/update";
  }

  get validationRules(): ValidationRule[] {
    return [];
  }
}

@MediatorHandler(UpdateUserCommandHandler.name)
export class UpdateUserCommandHandler implements IRequestHandler<UpdateUserCommand, User> {
  constructor(
    @Inject(HttpService) private _httpService: HttpService,
    @Inject(NotificationService) private _notificationService: NotificationService,
  ) {

  }
  async Handle(request: UpdateUserCommand) {
    let httpRequest = new HttpRequest(request.url,request);
    return await this._httpService.Put<UpdateUserCommand,ServiceResult<User>>(httpRequest).toPromise().then(res => {
      this._notificationService.showSuccessMessage()
      return res.objResult
    })
  }
}

File: Backend/Microservices/Admin/Eefa.Admin.Application/UseCases/User/Command/Update/UpdateUserCommand.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using AutoMapper;
using Eefa.Admin.Application.CommandQueries.User.Query.Get;
using Library.Common;
using Library.Interfaces;
using Library.Mappings;
using Library.Models;
using Library.Utility;
using MediatR;
using Microsoft.EntityFrameworkCore;


namespace Eefa.Admin.Application.CommandQueries.User.Command.Update
{
    public class UpdateUserCommand : CommandBase, IRequest<ServiceResult>, IMapFrom<Data.Databases.Entities.User>,
        ICommand
    {
        public int Id { get; set; }

        /// <summary>
        /// کد پرسنلی
        /// </summary>
        public int PersonId { get; set; } = default!;

        /// <summary>
        /// نام کاربری
        /// </summary>
        public string Username { get; set; } = default!;

        /// <summary>
        /// آیا قفل شده است؟
        /// </summary>
        public bool IsBlocked { get; set; } = default!;

        /// <summary>
        /// علت قفل شدن
        /// </summary>
        public int? BlockedReasonBaseId { get; set; }

        /// <summary>
        /// رمز موقت
        /// </summary>
        public string? OneTimePassword { get; set; }

        /// <summary>
        /// رمز
        /// </summary>
        public string Password { get; set; } = default!;
        public string ConfirmPassword { get; set; } = default!;
        public DateTime PasswordExpiryDate { get; set; }
        public ICollection<int> UserAllowedYears { get; set; }

        public IList<int> RolesIdList { get; set; }

        public void Mapping(Profile profile)
        {
            profile.CreateMap<UpdateUserCommand, Data.Databases.Entities.User>()
                .IgnoreAllNonExisting();
        }
    }


    public class UpdateUserCommandHandler : IRequestHandler<UpdateUserCommand, ServiceResult>
    {
        private readonly IRepository _repository;
        private readonly IMapper _mapper;
        private readonly IConfigurationAccessor _configurationAccessor;
        private readonly IMediator _mediator;

        public UpdateUserCommandHandler(IRepository repository, IMapper mapper,
            IConfigurationAccessor configurationAccessor, IMediator mediator)
        {
            _mapper = mapper;
            _configurationAccessor = configurationAccessor;
            _mediator = mediator;
            _repository = repository;
        }

        public async Task<ServiceResult> Handle(UpdateUserCommand request, CancellationToken cancellationToken)
        {
            var entity = await _repository
                .Find<Data.Databases.Entities.User>(c =>
                    c.ObjectId(request.Id)).Include(x => x.UserRoleUsers)
                .Include(x => x.UserYearUsers)
                .FirstOrDefaultAsync(cancellationToken);

            if (!string.IsNullOrEmpty(request.Password))
            {

                if (request.Password != request.ConfirmPassword)
                {
                    throw new Exception("InvalidConfirmPassword");
                }

                request.Password = Encryption.Symmetric.CreateMd5Hash(request.Password);


                entity.Password = request.Password;
                //entity.PasswordExpiryDate = request.PasswordExpiryDate;
            }

            entity.PersonId = request.PersonId;
            entity.Username = request.Username.ToLower();
            entity.IsBlocked = request.IsBlocked;
            if (request.BlockedReasonBaseId != null)
            {
                entity.BlockedReasonBaseId = request.BlockedReasonBaseId;
            }

            entity.OneTimePassword = request.OneTimePassword;
            
            foreach (var removedRole in entity.UserRoleUsers.Select(x => x.RoleId).Except(request.RolesIdList))
            {
                var deletingEntity = await _repository
                    .Find<Data.Databases.Entities.UserRole>(c =>
                        c.ConditionExpression(x => x.RoleId == removedRole && x.UserId == entity.Id))
                    .FirstOrDefaultAsync(cancellationToken);

                _repository.Delete(deletingEntity);
            }

            foreach (var addedRole in request.RolesIdList.Except(entity.UserRoleUsers.Select(x => x.RoleId)))
            {
                _repository.Insert(new Data.Databases.Entities.UserRole()
                {
                    UserId = entity.Id,
                    RoleId = addedRole,
                });
            }

            foreach (var removedyear in entity.UserYearUsers.Select(x => x.YearId).Except(request.UserAllowedYears))
            {
                var deletingEntity = await _repository
                    .Find<Data.Databases.Entities.UserYear>(c =>
                        c.ConditionExpression(x => x.YearId == removedyear && x.UserId == entity.Id))
                    .FirstOrDefaultAsync(cancellationToken);

                _repository.Delete(deletingEntity);
            }

            foreach (var addedYear in request.UserAllowedYears.Except(entity.UserYearUsers.Select(x => x.YearId)))
            {
                _repository.Insert(new Data.Databases.Entities.UserYear()
                {
                    UserId = entity.Id,
                    YearId = addedYear,
                });
            }


            var updateEntity = _repository.Update(entity);
            var res = await request.Save(_repository, updateEntity.Entity, cancellationToken);
            if (res.Succeed)
            {
                return await _mediator.Send(new GetUserQuery() { Id = updateEntity.Entity.Id }, cancellationToken);
            }
            else
            {
                return res;
            }
        }
    }
}